database/src/ops/block.rs

```diff
@@ -1,89 +1,413 @@
 //! Blocks.
 
 //---------------------------------------------------------------------------------------------------- Import
+use std::sync::Arc;
 
-//---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn add_block() {
-    todo!()
-}
+use bytemuck::TransparentWrapper;
+use curve25519_dalek::{constants::ED25519_BASEPOINT_POINT, Scalar};
+use monero_serai::{
+    block::Block,
+    transaction::{Input, Timelock, Transaction},
+};
 
-/// TODO
-pub fn add_block_data() {
-    todo!()
-}
+use cuprate_types::{ExtendedBlockHeader, TransactionVerificationData, VerifiedBlockInformation};
 
-/// TODO
-pub fn pop_block() {
-    todo!()
-}
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::{
+        blockchain::chain_height,
+        key_image::{add_key_image, remove_key_image},
+        macros::doc_error,
+        output::{
+            add_output, add_rct_output, get_rct_num_outputs, remove_output, remove_rct_output,
+        },
+        tx::{add_tx, get_num_tx, remove_tx},
+    },
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        AmountIndex, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, OutputFlags,
+        PreRctOutputId, RctOutput, TxHash,
+    },
+    StorableVec,
+};
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps/adds them to the appropriate database tables.
+///
+#[doc = doc_error!()]
+///
+/// # Panics
+/// This function will panic if:
+/// - `block.height > u32::MAX` (not normally possible)
+/// - `block.height` is not != [`chain_height`]
+///
+/// # Already exists
+/// This function will operate normally even if `block` already
+/// exists, i.e., this function will not return `Err` even if you
+/// call this function infinitely with the same block.
+// no inline, too big.
+#[allow(clippy::too_many_lines)]
+#[allow(clippy::manual_assert)] // assert doesn't let you `{}`
```

Comment:
I am not sure what this comment means, if it means you can do formatting with assert, I am pretty sure you can: https://doc.rust-lang.org/std/macro.assert.html#custom-messages